File: Game_Engine/Class/Camera.py
import pygame as pg
import pyautogui
from Class.Utils.matrix_functions import *

class Camera:
    def __init__(self, render, position):
        self.render = render
        self.position = np.array([*position, 1.0])
        self.forward = np.array([0, 0, 1, 1])
        self.up = np.array([0, 1, 0, 1])
        self.right = np.array([1, 0, 0, 1])
        self.h_fov = math.pi / 3
        self.v_fov = self.h_fov * (render.HEIGHT / render.WIDTH)
        self.near_plane = 0.1
        self.far_plane = 100
        self.moving_speed = 0.3
        self.rotation_speed = 0.015

        self.anglePitch = 0
        self.angleYaw = 0
        self.angleRoll = 0

        self.sensitivity = 0.1

    def control(self):
        key = pg.key.get_pressed()
        if key[pg.K_q]:
            self.position -= self.right * self.moving_speed
        if key[pg.K_d]:
            self.position += self.right * self.moving_speed
        if key[pg.K_z]:
            self.position += self.forward * self.moving_speed
        if key[pg.K_s]:
            self.position -= self.forward * self.moving_speed
        if key[pg.K_SPACE]:
            self.position += self.up * self.moving_speed
        if key[pg.K_LSHIFT]:
            self.position -= self.up * self.moving_speed

        if key[pg.K_RIGHT]:
            self.camera_yaw(-self.rotation_speed)
        if key[pg.K_LEFT]:
            self.camera_yaw(self.rotation_speed)
        if key[pg.K_DOWN]:
            self.camera_pitch(-self.rotation_speed)
        if key[pg.K_UP]:
            self.camera_pitch(self.rotation_speed)
        self.camera_mouse()
        
    
    def camera_mouse(self):
        # get mouse position
        print(pyautogui.position())
        # calc the difference between the center of the screen and the mouse position
        mouse_x, mouse_y = pyautogui.position()
        diff_x = self.render.WIDTH / 2 - mouse_x
        diff_y = self.render.HEIGHT / 2 - mouse_y
        # set the mouse position to the center of the screen
        if diff_x > 0:
            self.camera_yaw(self.rotation_speed * self.sensitivity * -diff_x)
        elif diff_x < 0:
            self.camera_yaw(-self.rotation_speed * self.sensitivity * diff_x)
        if diff_y > 0:
            self.camera_pitch(self.rotation_speed * self.sensitivity * -diff_y)
        elif diff_y < 0:
            self.camera_pitch(-self.rotation_speed * self.sensitivity * diff_y)
        # calc the angle of the mouse movement
        pg.mouse.set_pos(self.render.WIDTH / 2, self.render.HEIGHT / 2)

    def camera_yaw(self, angle):
        self.angleYaw += angle

    def camera_pitch(self, angle):
        self.anglePitch += angle

    def axiiIdentity(self):
        self.forward = np.array([0, 0, 1, 1])
        self.up = np.array([0, 1, 0, 1])
        self.right = np.array([1, 0, 0, 1])

    def camera_update_axii(self):
        # rotate = rotate_y(self.angleYaw) @ rotate_x(self.anglePitch)
        rotate = rotate_x(self.anglePitch) @ rotate_y(self.angleYaw)  # this concatenation gives right visual
        self.axiiIdentity()
        self.forward = self.forward @ rotate
        self.right = self.right @ rotate
        self.up = self.up @ rotate

    def camera_matrix(self):
        self.camera_update_axii()
        return self.translate_matrix() @ self.rotate_matrix()

    def translate_matrix(self):
        x, y, z, w = self.position
        return np.array([
            [1, 0, 0, 0],
            [0, 1, 0, 0],
            [0, 0, 1, 0],
            [-x, -y, -z, 1]
        ])

    def rotate_matrix(self):
        rx, ry, rz, w = self.right
        fx, fy, fz, w = self.forward
        ux, uy, uz, w = self.up
        return np.array([
            [rx, ux, fx, 0],
            [ry, uy, fy, 0],
            [rz, uz, fz, 0],
            [0, 0, 0, 1]
        ])

File: CPPMinecraft/src/engine/Camera.cpp
/*
** EPITECH PROJECT, 2022
** fun
** File description:
** Camera
*/

#include "Camera.hpp"

Camera::Camera(sf::Vector2i window_size, std::vector<float> position)
{
    position.push_back(1);
    this->_position = position;
    this->_forward = std::vector<float>{0, 0, 1, 1};
    this->_up = std::vector<float>{0, 1, 0, 1};
    this->_right = std::vector<float>{1, 0, 0, 1};
    this->_hfov = PI / 3;
    this->_window_size = window_size;
    this->_vfov = _hfov * (_window_size.y / _window_size.x);
    this->_near = 0.1;
    this->_far = 100;
    this->_movingSpeed = 0.3;
    this->_rotationSpeed = 0.015;
    this->_mouseSensitivity = 0.1;
    this->_yaw = 0;
    this->_pitch = 0;
    this->_roll = 0;
}

Camera::~Camera()
{
}

void Camera::setWindowSize(sf::Vector2i window_size) {
    this->_window_size = window_size;
    this->_vfov = _hfov * (_window_size.y / _window_size.x);
}

void Camera::control(sf::Event::KeyEvent key)
{
    if (key.code == sf::Keyboard::Z) {
        this->_position = matrix_add(this->_position, matrix_mult(this->_forward, this->_movingSpeed));
    }
    if (key.code == sf::Keyboard::S) {
        this->_position = matrix_sub(this->_position, matrix_mult(this->_forward, this->_movingSpeed));
    }
    if (key.code == sf::Keyboard::Q) {
        this->_position = matrix_sub(this->_position, matrix_mult(this->_right, this->_movingSpeed));
    }
    if (key.code == sf::Keyboard::D) {
        this->_position = matrix_add(this->_position, matrix_mult(this->_right, this->_movingSpeed));
    }
    if (key.code == sf::Keyboard::Space) {
        this->_position = matrix_add(this->_position, matrix_mult(this->_up, this->_movingSpeed));
    }
    if (key.code == sf::Keyboard::LShift) {
        this->_position = matrix_sub(this->_position, matrix_mult(this->_up, this->_movingSpeed));
    }
}

void Camera::cameraYaw(float angle)
{
    this->_yaw += angle;
}

void Camera::cameraPitch(float angle)
{
    this->_pitch += angle;
}

void Camera::axiiIdentity()
{
    this->_forward = std::vector<float>{0, 0, 1, 1};
    this->_up = std::vector<float>{0, 1, 0, 1};
    this->_right = std::vector<float>{1, 0, 0, 1};
}

void Camera::cameraUpdateAxii()
{
    std::vector<std::vector<float>> rotate = matrix_mult(rotateX(this->_pitch), rotateY(this->_yaw));
    this->axiiIdentity();
    this->_forward = matrix_mult(this->_forward, rotate);
    this->_right = matrix_mult(this->_right, rotate);
    this->_up = matrix_mult(this->_up, rotate);
}

std::vector<std::vector<float>> Camera::translateMatrix()
{
    return std::vector<std::vector<float>> {
        {1, 0, 0, 0},
        {0, 1, 0, 0},
        {0, 0, 1, 0},
        {-this->_position[0], -this->_position[1], -this->_position[2], 1}
    };
}

std::vector<std::vector<float>> Camera::rotateMatrix()
{
    return std::vector<std::vector<float>> {
        {this->_right[0], this->_up[0], this->_forward[0], 0},
        {this->_right[1], this->_up[1], this->_forward[1], 0},
        {this->_right[2], this->_up[2], this->_forward[2], 0},
        {0, 0, 0, 1}
    };
}

std::vector<std::vector<float>> Camera::cameraMatrix()
{
    this->cameraUpdateAxii();
    return matrix_mult(this->translateMatrix(), this->rotateMatrix());
}
